publish-mqtt/src/main.rs

```diff
@@ -309,51 +364,57 @@ async fn check_for_sensors(
     for props in sensors {
         if sensor_names.contains_key(&props.mac_address)
             && !state
-                .sensors_to_connect
+                .sensors
                 .iter()
-                .chain(state.sensors_connected.iter())
                 .find(|s| s.mac_address == props.mac_address)
                 .is_some()
         {
-            state
-                .sensors_to_connect
-                .push_back(Sensor::new(props, &sensor_names))
+            state.sensors.push(Sensor::new(props, &sensor_names))
         }
     }
     Ok(())
 }
 
-async fn connect_first_sensor_in_queue(
+async fn connect_sensor_at_idx(
+    state: Arc<Mutex<SensorState>>,
     session: &MijiaSession,
-    homie: &mut HomieDevice,
-    sensors_connected: &mut Vec<Sensor>,
-    sensors_to_connect: &mut VecDeque<Sensor>,
+    idx: usize,
 ) -> Result<(), anyhow::Error> {
-    println!(
-        "{} sensors in queue to connect, {} connected.",
-        sensors_to_connect.len(),
-        sensors_connected.len()
-    );
+    {
+        state.lock().await.sensors[idx].connection_status = ConnectionStatus::Connecting {
+            reserved_until: Instant::now() + Duration::from_secs(5 * 60),
+        }
+    }
     // Try to connect to a sensor.
-    if let Some(mut sensor) = sensors_to_connect.pop_front() {
-        println!("Trying to connect to {}", sensor.name);
-        match connect_start_sensor(session, homie, &mut sensor).await {
-            Err(e) => {
-                println!("Failed to connect to {}: {:?}", sensor.name, e);
-                sensors_to_connect.push_back(sensor);
-            }
-            Ok(()) => {
-                println!("Connected to {} and started notifications", sensor.name);
-                sensors_connected.push(sensor);
-            }
+    let mut sensor = clone_sensor_at_idx(state.clone(), idx).await;
+    println!("Trying to connect to {}", sensor.name);
+    let status = connect_start_sensor(session, &mut sensor).await;
```

Comment:
I'm a bit sad that connect_start_sensor() mutates sensor rather than just returning the new connection_status, but think let's merge this and I'll have a crack at it over the weekend.